components/scheduler/scheduler.go

```diff
@@ -0,0 +1,414 @@
+	// exactly, but it's just informational so it's close enough.
+	now := time.Now()
+
+	if !start.Manual {
+		// record metric only for _scheduled_ actions, not trigger/backfill, otherwise it's not meaningful
+		desiredTime := cmp.Or(start.DesiredTime, start.ActualTime)
+		s.metrics.Timer(metrics.ScheduleActionDelay.Name()).Record(now.Sub(desiredTime.AsTime()))
+	}
+
+	return &schedpb.ScheduleActionResult{
+		ScheduleTime: start.ActualTime,
+		ActualTime:   timestamppb.New(now),
+		StartWorkflowResult: &commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      res.RunId,
+		},
+	}, nil
+}
+
+func (s *Scheduler) terminateWorkflow(ex *commonpb.WorkflowExecution) {
+	// TODO: remove after https://github.com/temporalio/sdk-go/issues/1066
+	// TODO(Tianyu): hardcoded wait time
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	rreq := &historyservice.TerminateWorkflowExecutionRequest{
+		NamespaceId: s.Args.State.NamespaceId,
+		TerminateRequest: &workflowservice.TerminateWorkflowExecutionRequest{
+			Namespace: s.Args.State.Namespace,
+			// only set WorkflowId so we cancel the latest, but restricted by FirstExecutionRunId
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+			Reason:              "terminated by schedule overlap policy",
+			Identity:            s.identity(),
+			FirstExecutionRunId: ex.RunId,
+		},
+	}
+	_, err := s.historyClient.TerminateWorkflowExecution(ctx, rreq)
+
+	// TODO(Tianyu): original implementation translates this error which may not be relevant any more
+	if err != nil {
+		s.logger.Error("terminate workflow failed", tag.WorkflowID(ex.WorkflowId), tag.Error(err))
+		s.metrics.Counter(metrics.ScheduleTerminateWorkflowErrors.Name()).Record(1)
+	}
+}
+
+func (s *Scheduler) cancelWorkflow(ex *commonpb.WorkflowExecution) {
+	// TODO: remove after https://github.com/temporalio/sdk-go/issues/1066
+	// TODO(Tianyu): hardcoded wait time
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	rreq := &historyservice.RequestCancelWorkflowExecutionRequest{
+		NamespaceId: s.Args.State.NamespaceId,
+		CancelRequest: &workflowservice.RequestCancelWorkflowExecutionRequest{
+			Namespace: s.Args.State.Namespace,
+			// only set WorkflowId so we cancel the latest, but restricted by FirstExecutionRunId
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+			Identity:            s.identity(),
+			RequestId:           uuid.NewString(),
+			FirstExecutionRunId: ex.RunId,
+			Reason:              "cancelled by schedule overlap policy",
+		},
+	}
+	_, _ = s.historyClient.RequestCancelWorkflowExecution(ctx, rreq)
+	// Note: the local activity has completed (or failed) here but the workflow might take time
+	// to close since a cancel is only a request.
+	// If this failed, that's okay, we'll try it again the next time we try to take an action.
```

Comment:
not relevant

Reply:
The "local activity" part is not accurate but the rest of the comment is still relevant.